### PR TITLE
Added safety checks for BF KNN on GPU

### PR DESCRIPTION
**Pull request by amgrigoriev**

### CODE REVIEW

File: cpp/daal/src/sycl/select_indexed.cpp [hunk 1]
```diff
@@ -42,6 +45,12 @@ void run_quick_select_simd(ExecutionContextIface & context, ClKernelFactoryIface
                            uint32_t vectorSize, uint32_t lastVectorSize, uint32_t vectorOffset, QuickSelectIndexed::Result & result,
                            services::Status * status)
 {
+    if (nRndSeq <= maxInt32 && vectorSize <= maxInt32 && lastVectorSize <= maxInt32 && nK <= maxInt32 && vectorOffset <= maxInt32)
```
Comment (averbukh) [inline cpp/daal/src/sycl/select_indexed.cpp:48]: Do we need to check lower bound as well?

File: cpp/daal/src/algorithms/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i [hunk 2]
```diff
@@ -88,16 +91,26 @@ Status KNNClassificationPredictKernelUCAPI<algorithmFpType>::compute(const Numer
     const Parameter * const parameter = static_cast<const Parameter *>(par);
     const uint32_t k                  = parameter->k;
 
-    const size_t nQueryRows  = ntData->getNumberOfRows();
-    const size_t nLabelRows  = labels->getNumberOfRows();
-    const size_t nDataRows   = points->getNumberOfRows() < nLabelRows ? points->getNumberOfRows() : nLabelRows;
-    const uint32_t nFeatures = points->getNumberOfColumns();
+    if (ntData->getNumberOfRows() > maxInt32SizeT || labels->getNumberOfRows() > maxInt32SizeT)
```
Comment (rlnx) [inline cpp/daal/src/algorithms/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i:94]: Can this and the following checks be replaced by the `DAAL_CHECK` for consistency?

File: cpp/daal/src/algorithms/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i [hunk 3]
```diff
@@ -32,6 +33,9 @@
 
 #include "src/externals/service_ittnotify.h"
 
+const size_t maxInt32SizeT = static_cast<size_t>(daal::services::internal::MaxVal<int32_t>::get());
+const uint32_t maxInt32    = static_cast<uint32_t>(daal::services::internal::MaxVal<int32_t>::get());
```
Comment (averbukh) [inline cpp/daal/src/algorithms/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i:37]: Inconsistency between var names
Could you replace const to constexpr?

File: cpp/daal/src/algorithms/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i [hunk 4]
```diff
@@ -88,16 +91,21 @@ Status KNNClassificationPredictKernelUCAPI<algorithmFpType>::compute(const Numer
     const Parameter * const parameter = static_cast<const Parameter *>(par);
     const uint32_t k                  = parameter->k;
 
-    const size_t nQueryRows  = ntData->getNumberOfRows();
-    const size_t nLabelRows  = labels->getNumberOfRows();
-    const size_t nDataRows   = points->getNumberOfRows() < nLabelRows ? points->getNumberOfRows() : nLabelRows;
-    const uint32_t nFeatures = points->getNumberOfColumns();
+    DAAL_CHECK(ntData->getNumberOfRows() <= maxInt32SizeT && labels->getNumberOfRows() <= maxInt32SizeT,
+               services::ErrorIncorrectNumberOfRowsInInputNumericTable);
+    DAAL_CHECK(points->getNumberOfColumns() <= maxInt32SizeT, services::ErrorIncorrectNumberOfColumnsInInputNumericTable);
```
Comment (averbukh) [inline cpp/daal/src/algorithms/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i:96]: Could you avoid assumption that result of function call will not be different for check and value usage?

File: cpp/daal/src/sycl/select_indexed.cpp [hunk 5]
```diff
@@ -22,11 +22,14 @@
 #include "src/externals/service_rng.h"
 #include "src/algorithms/engines/engine_batch_impl.h"
 #include "services/daal_string.h"
+#include "src/services/service_data_utils.h"
 #include "src/externals/service_ittnotify.h"
 
 using namespace daal::data_management;
 using namespace daal::services::internal;
 
+const uint32_t maxInt32 = static_cast<uint32_t>(daal::services::internal::MaxVal<int32_t>::get());
```
Comment (averbukh) [inline cpp/daal/src/sycl/select_indexed.cpp:31]: constexpr?

File: cpp/daal/src/sycl/select_indexed.cpp [hunk 6]
```diff
@@ -42,6 +45,13 @@ void run_quick_select_simd(ExecutionContextIface & context, ClKernelFactoryIface
                            uint32_t vectorSize, uint32_t lastVectorSize, uint32_t vectorOffset, QuickSelectIndexed::Result & result,
                            services::Status * status)
 {
+    if (nRndSeq > maxInt32 || nRndSeq == 0 || vectorSize > maxInt32 || lastVectorSize > vectorSize || nK > vectorSize || nK > lastVectorSize
+        || vectorOffset > maxInt32)
+    {
+        services::internal::tryAssignStatus(status, Status(services::ErrorBufferSizeIntegerOverflow));
```
Comment (averbukh) [inline cpp/daal/src/sycl/select_indexed.cpp:51]: Are you sure that this status code is the best choice for every condition in 'if'? For instance, if nRndSeq == 0, is it really integer overflow for buffer size?

File: cpp/daal/src/algorithms/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i (hunk 3 above)
Comment (amgrigoriev) [reply cpp/daal/src/algorithms/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i:37]: constexpr can't be compiled

File: cpp/daal/src/sycl/select_indexed.cpp (hunk 5 above)
Comment (amgrigoriev) [reply cpp/daal/src/sycl/select_indexed.cpp:31]: Can't be compiled

File: cpp/daal/src/algorithms/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i (hunk 3 above)
Comment (amgrigoriev) [reply cpp/daal/src/algorithms/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i]: Var name was changed

File: cpp/daal/src/algorithms/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i [hunk 7]
```diff
@@ -88,16 +93,28 @@ Status KNNClassificationPredictKernelUCAPI<algorithmFpType>::compute(const Numer
...
+    const size_t nQueryRowsSizeT     = ntData->getNumberOfRows();
+    const size_t nQueryFeaturesSizeT = ntData->getNumberOfColumns();
+    const size_t nLabelRowsSizeT     = labels->getNumberOfRows();
+    const size_t nDataRowsSizeT      = points->getNumberOfRows();
+    const size_t nTrainFeaturesSizeT = points->getNumberOfColumns();
+
+    DAAL_CHECK(nQueryRowsSizeT <= maxInt32AsSizeT && nLabelRowsSizeT <= maxInt32AsSizeT && nDataRowsSizeT <= maxInt32AsSizeT,
+               services::ErrorIncorrectNumberOfRowsInInputNumericTable);
+    DAAL_CHECK(nTrainFeaturesSizeT <= maxInt32AsSizeT && nTrainFeaturesSizeT == nQueryFeaturesSizeT,
+               services::ErrorIncorrectNumberOfColumnsInInputNumericTable);
```
Comment (KulikovNikita) [inline cpp/daal/src/algorithms/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i:105]:
```suggestion
    DAAL_CHECK(nQueryRowsSizeT <= maxInt32AsSizeT, services::ErrorIncorrectNumberOfRowsInInputNumericTable);
    DAAL_CHECK(nLabelRowsSizeT <= maxInt32AsSizeT, services::ErrorIncorrectNumberOfRowsInInputNumericTable); 
    DAAL_CHECK(nDataRowsSizeT <= maxInt32AsSizeT, services::ErrorIncorrectNumberOfRowsInInputNumericTable)
    DAAL_CHECK(nTrainFeaturesSizeT <= maxInt32AsSizeT, services::ErrorIncorrectNumberOfColumnsInInputNumericTable);
    DAAL_CHECK(nTrainFeaturesSizeT == nQueryFeaturesSizeT, services::ErrorIncorrectNumberOfColumnsInInputNumericTable);
```

File: cpp/daal/src/algorithms/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i [hunk 8]
```diff
@@ -88,16 +93,28 @@ Status KNNClassificationPredictKernelUCAPI<algorithmFpType>::compute(const Numer
     const Parameter * const parameter = static_cast<const Parameter *>(par);
     const uint32_t k                  = parameter->k;
```
Comment (KulikovNikita) [inline cpp/daal/src/algorithms/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i:94]: Should it be checked for overflow?

File: cpp/daal/src/algorithms/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i [hunk 9]
```diff
@@ -190,17 +208,14 @@ Status KNNClassificationPredictKernelUCAPI<algorithmFpType>::compute(const Numer
...
+services::Status KNNClassificationPredictKernelUCAPI<algorithmFpType>::copyPartialDistancesAndLabels(
     ExecutionContextIface & context, const UniversalBuffer & distances, const UniversalBuffer & labels, UniversalBuffer & partialDistances,
     UniversalBuffer & partialLabels, uint32_t queryBlockRows, uint32_t k, uint32_t nChunk, uint32_t totalNumberOfChunks)
 {
     DAAL_ITTNOTIFY_SCOPED_TASK(compute.copyPartialSelections);
-    if (k > static_cast<uint32_t>(INT_MAX) || totalNumberOfChunks > static_cast<uint32_t>(INT_MAX) || nChunk > static_cast<uint32_t>(INT_MAX))
-    {
-        return services::Status(services::ErrorBufferSizeIntegerOverflow);
-    }
+    DAAL_CHECK(totalNumberOfChunks <= maxInt32AsUint32T && nChunk <= maxInt32AsUint32T, services::ErrorBufferSizeIntegerOverflow);
```
Comment (KulikovNikita) [inline cpp/daal/src/algorithms/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i:216]:
```suggestion
    DAAL_CHECK(totalNumberOfChunks <= maxInt32AsUint32T, services::ErrorBufferSizeIntegerOverflow);
    DAAL_CHECK(nChunk <= maxInt32AsUint32T, services::ErrorBufferSizeIntegerOverflow);
```

File: cpp/daal/src/sycl/select_indexed.cpp [hunk 10]
```diff
@@ -39,63 +42,77 @@ namespace selection
...
+                                       QuickSelectIndexed::Result & result)
 {
-    auto func_kernel = kernelFactory.getKernel("quick_select_group", status);
-    DAAL_CHECK_STATUS_PTR(status);
+    services::Status status;
+    if (nRndSeq > maxInt32 || nRndSeq == 0 || vectorSize > maxInt32 || lastVectorSize > vectorSize || nK > vectorSize || nK > lastVectorSize
+        || vectorOffset > maxInt32)
+    {
+        return Status(services::ErrorBufferSizeIntegerOverflow);
+    }
```
Comment (KulikovNikita) [inline cpp/daal/src/sycl/select_indexed.cpp:55]: Shouldn't it been done with `DAAL_CHECK`?

**Review by KulikovNikita — APPROVED** [review]

File: cpp/daal/src/sycl/select_indexed.cpp [hunk 11]
```diff
@@ -22,11 +22,15 @@
...
 #include "src/algorithms/engines/engine_batch_impl.h"
 #include "services/daal_string.h"
+#include "src/services/service_data_utils.h"
 #include "src/externals/service_ittnotify.h"
+#include "services/internal/sycl/daal_defines_sycl.h"
 
 using namespace daal::data_management;
 using namespace daal::services::internal;
 
+constexpr size_t maxInt32AsUint32T = static_cast<uint32_t>(daal::services::internal::MaxVal<int32_t>::get());
```
Comment (KulikovNikita) [inline cpp/daal/src/sycl/select_indexed.cpp:32]: Looks incorrect since it is actually have `size_t` type.

**Review by KulikovNikita — APPROVED** [review]

File: cpp/daal/src/algorithms/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i [hunk 12]
```diff
@@ -48,6 +53,7 @@ using sort::RadixSort;
 using selection::QuickSelectIndexed;
 using selection::SelectIndexed;
 using selection::SelectIndexedFactory;
+;
```
Comment (vepifanov) [inline cpp/daal/src/algorithms/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i:56]: Extra semicolon can be erased.

**Review by vepifanov — APPROVED** [review]

**Pull request merged**